pennylane/pauli/dla/khk.py

```diff
@@ -0,0 +1,564 @@
+    U, S, _ = np.linalg.svd(projections.T)
+
+    # Choose columns of U corresponding to non-zero singular values
+    rank = np.sum(S > tol)
+    basis = U[:, :rank]
+
+    return basis.T  # Transpose to get row vectors
+
+
+def linearly_independent_set_svd(vectors):
+    """Find a basis given a set of vectors"""
+    A = np.array(vectors).T  # Transpose to get vectors as columns
+
+    # Perform Singular Value Decomposition
+    U, S, VT = np.linalg.svd(A)
+
+    # Determine the rank of the matrix (number of non-zero singular values)
+    rank = np.sum(S > 1e-10)  # Use a threshold to determine non-zero singular values
+
+    # Select the first 'rank' columns of U (corresponding to non-zero singular values)
+    independent_vectors = U[:, :rank]
+
+    return independent_vectors.T  # Transpose back to original orientation
+
+
+def compute_csa_new(g, m, ad, which=0, tol=1e-14, verbose=0):
+    r"""
+    Compute cartan subalgebra of g given m, the odd parity subspace. I.e. the maximal Abelian subalgebra in m.
+
+    Args:
+        g (List[PauliSentence]): DLA in terms of PauliSentence instnaces
+        m (List[PauliSentence]): Odd parity subspace :math:`\mathfrak{m}` from Cartan decomposition g = k + m
+        ad (Array): The ``(|g|, |g|, |g|)`` dimensional adjoint representation of g
+
+    Returns:
+        mtilde (List[PauliSentence]): the remaining operators in m, i.e. :math:`\mathfrak{m} = \tilde{\mathfrak{m}} \oplus \mathfrak{h}`
+        h (List[PauliSentence]): the Cartan subalgebra :math:`\mathfrak{h}`
+    """
+    # brute force translation from implementation of therooler in https://github.com/therooler/homogeneous_spaces/blob/main/A%20zoo%20of%20quantum%20gates.ipynb
+    # not at all optimized, some redundancies in re-computing adjoint representations
+    m = m.copy()
+
+    # all_pure_words = all(len(op)==1 for op in g) and all(len(op)==1 for op in m)
+    # if all_pure_words:
+    #     return _compute_csa_words(m, which)
+
+    g_vspace = PauliVSpace(g)
+    np_m = np.array([project(mop, g_vspace) for mop in m]).T
+    np_h = [project(m[which], g_vspace)]
+
+    iteration = 1
+    while True:
+        print(f"iteration: {iteration}") if verbose else None
+        kernel_intersection = np_m
+        for h_i in np_h:
+
+            # obtain adjoint rep of candidate h_i
+            adjoint_of_h_i = np.einsum("gab,a->gb", ad, h_i)
+            # compute kernel of adjoint
+            new_kernel = null_space(adjoint_of_h_i)
+
+            # intersect kernel to stay in m
+            kernel_intersection = _intersect_bases(kernel_intersection, new_kernel)
+
+        if kernel_intersection.shape[1] == len(np_h):
+            # No new vector was added from all the kernels
+            break
+
+        kernel_intersection = gram_schmidt(kernel_intersection)  # orthogonalize
+        for vec in kernel_intersection.T:
+            if is_independent(vec, np.array(np_h).T, tol):
+                np_h.append(vec)
+                break
```

Comment:
Could the problem be that in the case of sums of Paulis there is more than one new linearly independent vector, that we are not adding here? @dwierichs

Reply:
The idea is that even if there are multiple linearly independent objects at this point, we only add one and recompute which ones still commute with all the previous objects and the newly added one.
So this should be fine, structurally :thinking: